### PR TITLE
feat: send library version outside of compressed body as a debug signal

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

We add the library version to event properties. This is really useful when debugging problems. 

When decompression of the request body fails we get an error report in Sentry. But we can't add the library version to the Sentry report because it is inside the compressed content. 

This change adds the lib version to an HTTP header so that it would always be available in Sentry. We could (optionally) follow up by having the backend add that library version as a tag to Sentry so we get richer behaviour in the Sentry UI

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [x] Accounted for the impact of any changes across different browsers
- [x] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

**Comment by mariusandra** [comment]
There's already a standard way to describe who's making a request: the `user-agent` header. This is also what [one of our competitors](https://github.com/segmentio/analytics-node/blob/master/index.js#L276-L279) does.

I'd suggest following a similar pattern, since we don't have just one "Lib" either.

**Comment by mariusandra** [comment]
I think I need to do a 🤦 here. When reviewing this now, I started to wonder if setting a user agent in a browser is something that's done... and if some adblockers might consider that a strong signal that the user is up to something malicious or not. Apparently doing this [was disallowed](https://stackoverflow.com/a/42815264) not so long ago... so it could be penalised, tracked or just flakey yet again.

The [link I copied in my previous comment](https://github.com/segmentio/analytics-node/blob/master/index.js#L276-L279) shows actually the exact opposite of what I said in the comment itself. It's for the `-node` library and writes `// Don't set the user agent if we're on a browser.`

🤦 

How would you like to proceed? 😅 😢

**Comment by pauldambra** [comment]
Hmmm 🤔 

The nice thing about user-agent is that it's sort-of semantically correct. How about `posthog-user-agent`?

**Comment by mariusandra** [comment]
I still think:

1) This `posthog-user-agent` header makes a lot of sense from a practical perspective, and does indeed give us more debug help.
2) I'm scared this will allow adblockers to surgically disable all posthog-js requests, even if it's a first-party request to a self-hosted instance.

So... I'm still on the fence. Better be cautious? Could perhaps a ?v=1.2.3 style GET param be less obviously taggable while providing the same debug info?

Tagging a few people for a second opinion: @timgl @macobo @Twixes @guidoiaquinti

**Comment by macobo** [comment]
There are a lot of things that make me paranoid about this change:

1. Old browser support
2. Whether this affects CORS
3. Whether this changes the way compression works/xhr
4. Whether it increases the chance of getting blocked

I'd be cautiously for either a GET parameter or testing the heck out of this in a real-world setting.

**Comment by guidoiaquinti** [comment]
This is a bit outside my area of expertise. From a technical perspective, the user-agent is the solution I would use. From a business perspective I'm not sure it is the right solution.

**Comment by timgl** [comment]
I think Karl's points are valid. The simplest way of verifying whether this works is to make this a config option in posthog-js and set that config on our own website/app to start with. Then in a week or so we can see if we dropped events from any browser and if we're seeing any other errors.

File: src/send-request.js [hunk 1]
```diff
@@ -1,4 +1,19 @@
 import { _, console } from './utils'
+import Config from './config'
+
+export const addParamsToURL = (url, urlQueryArgs, parameterOptions) => {
+    console.log(parameterOptions, 'poo')
```
Comment (mariusandra) [inline src/send-request.js:5]: log `poo`

File: cypress/integration/capture.spec.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
 /// <reference types="cypress" />
+import { version } from '../../package.json'
 
 import { getBase64EncodedPayload, getGzipEncodedPayload, getLZStringEncodedPayload } from '../support/compression'
 
+const urlWithLibraryAndVersion = new RegExp(`&l=web&v=${version}`)
```
Comment (mariusandra) [inline cypress/integration/capture.spec.js:6]: I think `l=web&v=` is a pretty uniquely matchable pattern for aggressive traffic blockers, so perhaps we should infer `web` by default if only `v` is present? Other integrations than posthog-js can send it if they want, but there are benefits to not exposing it here.

File: cypress/integration/capture.spec.js [hunk 3]
```diff
@@ -198,14 +201,21 @@ describe('Event capture', () => {
 
     describe('capturing performance', () => {
         it('sends performance timing with the pageview when enabled', () => {
-            given('options', () => ({ capture_pageview: true, _capture_performance: true }))
+            given('options', () => ({
+                capture_pageview: true,
+                _capture_performance: true,
+                _library_and_version_in_query_params: true,
```
Comment (mariusandra) [inline cypress/integration/capture.spec.js:207]: I don't think this needs to be an option you configure. Just turn it on.

**Review by mariusandra — CHANGES_REQUESTED** [review]
some suggestions

**Review by mariusandra — APPROVED** [review]

**Pull request merged**